src/DataStructures/VectorImpl.hpp

```diff
@@ -639,15 +641,96 @@ template <typename T>
 constexpr bool is_derived_of_vector_impl_v =
     is_derived_of_vector_impl<T>::value;
 
+// impose strict equality for derived classes of VectorImpl; note that this
+// overrides intended behavior in blaze for comparison operators to use
+// approximate equality in favor of equality between containers being
+// appropriately recursive. This form primarily works by using templates to
+// ensure that our comparison operator is resolved with higher priority than the
+// blaze form as of blaze 3.8
+template <
+    typename Lhs, typename Rhs,
+    Requires<(is_derived_of_vector_impl_v<Lhs> or
+              is_derived_of_vector_impl_v<
+                  Rhs>)and not(std::is_base_of_v<blaze::Computation, Lhs> or
+                               std::is_base_of_v<blaze::Computation, Rhs>) and
+             not(std::is_same_v<Rhs, typename Lhs::ElementType> or
+                 std::is_same_v<Lhs, typename Rhs::ElementType>)> = nullptr>
+bool operator==(const Lhs& lhs, const Rhs& rhs) noexcept {
+  return blaze::equal<blaze::strict>(lhs, rhs);
+}
+
+template <
+    typename Lhs, typename Rhs,
+    Requires<(is_derived_of_vector_impl_v<Lhs> or
+              is_derived_of_vector_impl_v<
+                  Rhs>)and not(std::is_base_of_v<blaze::Computation, Lhs> or
+                               std::is_base_of_v<blaze::Computation, Rhs>) and
+             not(std::is_same_v<Rhs, typename Lhs::ElementType> or
+                 std::is_same_v<Lhs, typename Lhs::ElementType>)> = nullptr>
+bool operator!=(const Lhs& lhs, const Rhs& rhs) noexcept {
+  return not(lhs == rhs);
+}
+
+// Impose strict equality for any expression templates; note that
+// this overrides intended behavior in blaze for comparison
+// operators to use approximate equality in favor of equality
+// between containers being appropriately recursive. This form
+// primarily works by using templates to ensure that our
+// comparison operator is resolved with higher priority than the
+// blaze form as of blaze 3.8
+template <typename Lhs, typename Rhs,
+          Requires<std::is_base_of_v<blaze::Computation, Lhs> or
+                   std::is_base_of_v<blaze::Computation, Rhs>> = nullptr>
+bool operator==(const Lhs& lhs, const Rhs& rhs) noexcept {
+  return blaze::equal<blaze::strict>(lhs, rhs);
```

Comment:
since this implementation is the same as the one above, can you remove the `blaze::Computation` check? Or does that move the priority further down somehow?

Reply:
This is necessary for the overload priority to use this instead of the blaze ones -- `blaze::Computation` is a direct base of all of the expression templates, so uses this operator with higher priority than the conversions to `DenseVector`. However, `DenseVector` happens with the same priority as the other overload (to `VectorImpl`), so without this version with `blaze::Computation`, expression templates in comparison operators fail to compile with ambiguous overload.